### PR TITLE
L2 BLOB

### PR DESCRIPTION
**Pull request by qzhodl**
Added several small sections when comparing with the [spec](https://github.com/ethereum-optimism/specs/pull/383):
 - Purpose & Summary
 - Alternatives Considered
 - Reference Implementation

### CODE REVIEW

File: protocol/l2-blob.md [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+## Enabling BLOB Transactions in L2 EL
+
+The interface and implementation of L2 EL should remain consistent with L1 EL to ensure seamless migration of
+applications. Note that while BLOBs are gossiped within the L1 P2P network, for enshrined BLOB DA support in L2, the
+BLOBs should be sent directly to the L2 sequencer.
+
+## Uploading BLOB to Alt-DA
+
+The sequencer is responsible for uploading BLOBs to the Alt-DA layer. When the
+CL (op-node) receives the payload from EL via the engine API, it should inspect the envelope for any `BlobsBundle`
```
Comment (qizhou) [inline protocol/l2-blob.md]:
```suggestion
The sequencer is responsible for uploading BLOBs to the Alt-DA layer. When the
CL (op-node) of the sequencer receives the payload from its EL via the engine API, it should inspect the envelope for any `BlobsBundle`
```

File: protocol/l2-blob.md [hunk 2]
```diff
@@ -0,0 +1,199 @@
+# Purpose
+
+<!-- This section is also sometimes called “Motivations” or “Goals”. -->
+
+<!-- It is fine to remove this section from the final document,
+but understanding the purpose of the doc when writing is very helpful. -->
+Currently, EIP-4844 BLOB transactions are disabled on the OP Stack because Layer 2 (L2) BLOBs offer no advantage over L2 calldata when both use the same L1 DA. This forces L3s to rely on the same DA as their L2, unless they want to integrate third-party DA providers or manage the security risks associated with DA bridges. However, different DA solutions have their own advantages—L1 DA provides better security, while [Alt-DA](https://github.com/ethereum-optimism/specs/blob/main/specs/experimental/alt-da.md) offers lower costs.
```
Comment (blockchaindevsh) [inline protocol/l2-blob.md]: > This forces L3s to rely on the same DA as their L2

This statement is a bit confusing for me, L3s always have two choices: 
1. use the DA provided by L2.
2. use their own DA(L3 DA).

File: protocol/l2-blob.md [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+The purpose of this design document is to propose supporting L2 BLOBs based on Alt-DA, allowing us to leverage the advantages of both L1 DA and Alt-DA, creating a hybrid DA solution in the OP Stack.
+
+# Summary
+
+<!-- Most (if not all) documents should have a summary.
+While the length will likely be proportional to the length of the full document,
+the summary should be as succinct as possible. -->
+
+This document proposes support for BLOB transactions, enabling a **hybrid DA model** in the OP Stack:
```
Comment (blockchaindevsh) [inline protocol/l2-blob.md:20]: Besides **hybrid DA model**, is it worth to also mention that supporting L2 BLOBs reduces migration cost(techinical cost) for applications based on L1 BLOBs ?

File: protocol/l2-blob.md [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+<!-- A high level overview of the proposed solution.
+When there are multiple alternatives there should be an explanation
+of why one solution was picked over other solutions.
+As a rule of thumb, including code snippets (except for defining an external API)
+is likely too low level. -->
+
+## Enabling BLOB Transactions in L2 EL
+
+The interface and implementation of L2 EL should remain consistent with L1 EL to ensure seamless migration of
+applications. Note that while BLOBs are gossiped within the L1 P2P network, for enshrined BLOB DA support in L2, the
```
Comment (blockchaindevsh) [inline protocol/l2-blob.md]: > Note that while BLOBs are gossiped within the L1 P2P network, for enshrined BLOB DA support in L2, the BLOBs should be sent directly to the L2 sequencer.

The BLOB tx can also be sent to a relay node which forwards to the L2 sequencer.

File: protocol/l2-blob.md [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+BLOBs should be sent directly to the L2 sequencer.
+
+## Uploading BLOB to Alt-DA
+
+The sequencer is responsible for uploading BLOBs to the Alt-DA layer. When the
+CL (op-node) of the sequencer receives the payload from its EL via the engine API, it should inspect the envelope for any `BlobsBundle`
+and upload them to the Alt-DA. Only after ensuring successful BLOB uploads can the sequencer upload the block data
```
Comment (blockchaindevsh) [inline protocol/l2-blob.md:90]: > Only after ensuring successful BLOB uploads can the sequencer upload the block data to the L1 DA.

Sequencer is not responsible to upload block data to L1 DA, it's the batcher's responsibility.

File: protocol/l2-blob.md [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+## Enabling BLOB Transactions in L2 EL
+
+The interface and implementation of L2 EL should remain consistent with L1 EL to ensure seamless migration of
+applications. Note that while BLOBs are gossiped within the L1 P2P network, for enshrined BLOB DA support in L2, the
+BLOBs should be sent directly to the L2 sequencer.
+
+## Uploading BLOB to Alt-DA
+
+The sequencer is responsible for uploading BLOBs to the Alt-DA layer. When the
+CL (op-node) of the sequencer receives the payload from its EL via the engine API, it should inspect the envelope for any `BlobsBundle`
```
Comment (blockchaindevsh) [inline protocol/l2-blob.md:89]: > the CL (op-node) of the sequencer

From my understanding, sequencer is an optional component of CL.

File: protocol/l2-blob.md [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+1. When the op-node derives a batch from the `BatchQueue`, it collects all the BLOB commitments included in the batch.
+2. It tracks the challenge and resolution status for the collected BLOB hashes. If some BLOBs are challenged and fail
+to be resolved, the op-node should trigger a `ResetError` to initiate a reorg.
+3. If a reorg is triggered, the op-node will revert to an older block and rederive the chain. Upon encountering a
+block containing the expired BLOB hash, it will remove the corresponding transaction and pass the remaining block
+data to the EL.
+
+## Fault Proof
+
+The derivation pipeline integrates with fault proofs by adding additional hint types to the preimage oracle to query
```
Comment (blockchaindevsh) [inline protocol/l2-blob.md]: Why is it necessary? L2 state should have no dependency on BLOB, only BLOB hash.

File: protocol/l2-blob.md (hunk 5 above)
Comment (qzhodl) [reply protocol/l2-blob.md:90]: According to the OP [spec](https://github.com/ethereum-optimism/specs/blob/main/specs/glossary.md#sequencer), `The sequencer is a privileged actor, which receives L2 transactions from L2 users, creates L2 blocks using them, which it then submits to data availability provider (via a batcher). It also submits output roots to L1.` So IIUC, a sequence includes full node + batcher + proposer.

Comment (zhiqiangxu) [reply protocol/l2-blob.md:90]: OK, it's a more general concept.

Comment (blockchaindevsh) [reply protocol/l2-blob.md:90]: OK, it's a more general concept.

**Review by blockchaindevsh — APPROVED** [review]

File: protocol/l2-blob.md [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+transactions to Layer 1 (L1). This innovation has also enabled a variety of additional applications based on
+the BLOBs due to their low cost, such as [blob.fm](https://blob.fm/), [EthStorage](https://ethstorage.io), and
+[Ethscriptions](https://ethscriptions.com/). However, while the data upload costs have decreased, the execution
+costs on L1 remain high compared to L2, leading to high costs for L2 state proposals and non-financial applications
+that rely on BLOBs.
+
+The goal of this specification is to **support L2 BLOB transactions in the OP Stack**. This would allow L3 solutions,
+which settle on L2, to have an enshrined 4844-compatible DA layer that they can use directly, without needing to
+integrate third-party DA providers or deal with the security risks associated with DA bridges. Additionally, the
+applications mentioned above could migrate to L2 with minimal costs.
```
Comment (qizhou) [inline protocol/l2-blob.md]:
```suggestion
integrate third-party DA providers or deal with the security risks associated with DA bridges. Additionally, the
L1 applications mentioned above could migrate to L2 with minimal costs.
```

File: protocol/l2-blob.md [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+The sequencer is responsible for uploading BLOBs to the Alt-DA layer. When the
+CL (op-node) of the sequencer receives the payload from its EL via the engine API, it should inspect the envelope for any `BlobsBundle`
+and upload them to the Alt-DA. Only after ensuring successful BLOB uploads can the sequencer upload the block data
+to the L1 DA. Similarly, the sequencer may need to respond to any data availability challenges afterward.
+
+## Data Availability Challenge
+
+Any third party, including full nodes deriving L2 data, might find they cannot access the BLOB corresponding to
+the hash included in the BLOB transaction. In this case, they can initiate a data availability challenge. The
```
Comment (qizhou) [inline protocol/l2-blob.md]:
```suggestion
Any third party, including full nodes deriving L2 data, might find they cannot access the BLOB corresponding to
the hash included in the L2 BLOB transaction. In this case, they can initiate a data availability challenge on L1. The
```

**Review by qizhou — APPROVED** [review]

**Pull request merged**